Fix several issues with beatmap online ID management

bdach opened the pull request:
RFC.

This is an all-in-one diff with assorted changes after today's daily challenge fiasco. Caution, wall of text incoming.

Taking it commit by commit:

## [Adjust test to match desired reality](https://github.com/ppy/osu/commit/40c2d4e942453b583ff3dc41368992b347a474dd)

Self-explanatory.

## [Remove problematic online ID check](https://github.com/ppy/osu/commit/1a2e323c11158aab0433f144d1595e06f7a1fe20)

Would "close" https://github.com/ppy/osu/issues/27332 by essentially throwing up hands and deciding that it's probably fine not to check the garbage IDs in the `.osu` file.

## [Only use MD5 when performing metadata lookups](https://github.com/ppy/osu/commit/776fabd77c5a6225e69a0b14e79b9e097692320c)

Both online and offline using the cache.

The rationale behind this change is that after 1a2e323c11158aab0433f144d1595e06f7a1fe20, `TestPartiallyMaliciousSet()` fails in a way that cannot be reconciled without this sort of change.

The test exercises a scenario where the beatmap being imported has an online ID in the `.osu` file, but its hash does not match the online hash of the beatmap. This turns out to be a more frequent scenario than envisioned because of users doing stupid things with manual file editing rather than reporting issues properly.

The scenario is realistic only because the behaviour of the endpoint responsible for looking up beatmaps is such that if multiple parameters are given (e.g. all three of beatmap MD5, online ID, and filename), [it will try the three in succession](https://github.com/ppy/osu-web/blob/f6b341813be270de59ec8f9698428aa6422bc8ae/app/Http/Controllers/BeatmapsController.php#L260-L266), and the local metadata cache implementation reflected this implementation.

Because online ID and filename are inherently unreliable in this scenario due to being directly manipulable by clueless or malicious users, neither should not be used as a fallback.

## [Remove no longer applicable test](https://github.com/ppy/osu/commit/2c2f307a63c1c4eee8ccbc0bc5d339ccaaf308af)

After https://github.com/ppy/osu/pull/30453/commits/2c2f307a63c1c4eee8ccbc0bc5d339ccaaf308af the behaviour set up on the mock in the test in question is no longer realistic. Online metadata lookups will no longer fall back to online ID or filename.

## [Prefer not deleted models when picking model instances for reuse when importing](https://github.com/ppy/osu/commit/0e52797f2948aa60c6b375f80239c091d1a3fb7a)

This fell out while investigating why the issue with online IDs mismatching in the `.osu` could be worked around by importing the map three times in total when starting from it not being available locally.

Here follows an explanation of why that "helped".

Import 1:
- The beatmap set is imported normally.
- Online metadata population sees the online ID mismatch and resets it on the problematic beatmap.

Import 2:
- The existing beatmap set is found, but deemed not reusable because of the single beatmap having its ID reset to -1.
- The existing beatmap set is marked deleted, and all the IDs of its beatmaps are reset to -1.
- The beatmap set is reimported afresh.
- Online metadata population still sees the online ID mismatch and resets it on the problematic beatmap.

Note that at this point the first import *is still physically present in the database* but marked deleted.

Import 3:
- When trying to find the existing beatmap set to see if it can be reused, *the one pending deletion and with its IDs reset - the remnant from import 1 - is returned*.
- Because of this, `validateOnlineIds()` resets online IDs *on the model representing the current reimport*.
- The beatmap set is reimported yet again.
- With the online ID reset, the online metadata population check for online ID mismatch does not run because *the IDs were reset to -1* earlier.

Preferring undeleted models when picking the model instance for reuse prevents this scenario.

bdach committed:
Adjust test to match desired reality

bdach committed:
Remove problematic online ID check

bdach committed:
Only use MD5 when performing metadata lookups

Both online and offline using the cache.

The rationale behind this change is that in the current state of
affairs, `TestPartiallyMaliciousSet()` fails in a way that cannot be
reconciled without this sort of change.

The test exercises a scenario where the beatmap being imported has an
online ID in the `.osu` file, but its hash does not match the online
hash of the beatmap. This turns out to be a more frequent scenario than
envisioned because of users doing stupid things with manual file editing
rather than reporting issues properly.

The scenario is realistic only because the behaviour of the endpoint
responsible for looking up beatmaps is such that if multiple parameters
are given (e.g. all three of beatmap MD5, online ID, and filename), it
will try the three in succession:

	https://github.com/ppy/osu-web/blob/f6b341813be270de59ec8f9698428aa6422bc8ae/app/Http/Controllers/BeatmapsController.php#L260-L266

and the local metadata cache implementation reflected this
implementation.

Because online ID and filename are inherently unreliable in this
scenario due to being directly manipulable by clueless or malicious
users, neither should not be used as a fallback.

bdach committed:
Remove no longer applicable test

After dd06dd0e699311494412e36bc3f37bb055a01477 the behaviour set up on
the mock in the test in question is no longer realistic. Online
metadata lookups will no longer fall back to online ID or filename.

bdach committed:
Prefer not deleted models when picking model instances for reuse when importing

This fell out while investigating why the issue with online IDs
mismatching in the `.osu` could be worked around by importing the map
three times in total when starting from it not being available locally.

Here follows an explanation of why that "helped".

Import 1:
- The beatmap set is imported normally.
- Online metadata population sees the online ID mismatch and resets it
  on the problematic beatmap.

Import 2:
- The existing beatmap set is found, but deemed not reusable
  because of the single beatmap having its ID reset to -1.
- The existing beatmap set is marked deleted, and all the IDs of
  its beatmaps are reset to -1.
- The beatmap set is reimported afresh.
- Online metadata population still sees the online ID mismatch
  and resets it on the problematic beatmap.

Note that at this point the first import *is still physically present
in the database* but marked deleted.

Import 3:
- When trying to find the existing beatmap set to see if it can be
  reused, *the one pending deletion and with its IDs reset -
  - the remnant from import 1 - is returned*.
- Because of this, `validateOnlineIds()` resets online IDs
  *on the model representing the current reimport*.
- The beatmap set is reimported yet again.
- With the online ID reset, the online metadata population check for
  online ID mismatch does not run because *the IDs were reset to -1*
  earlier.

Preferring undeleted models when picking the model instance for reuse
prevents this scenario.

peppy commented on osu.Game/Beatmaps/BeatmapUpdaterMetadataLookup.cs:
```diff
@@ -85,11 +85,11 @@ public void Update(BeatmapSetInfo beatmapSet, bool preferOnlineFetch)
+            // this means that the online ID in the `.osu` file is not reliable, and this cannot be fixed server-side
+            // because updating the maps retroactively would break stable (by losing all of users' local scores).
 
             if (beatmapInfo.OnlineID == -1 && result.MD5Hash != beatmapInfo.MD5Hash)
```
On a quick read of the code, I'm not sure that we can ever hit this scenario anymore. The code run prior to this is a `tryLookup` which is based on md5 matching alone, so at this point it should be guaranteed that the MD5 matches.

I may be wrong, but we can probably do away with this method altogether?

bdach replied:
Good point, and the assertion holds up as far as I can tell. Removed in https://github.com/ppy/osu/commit/2b0fd3558f2f7f1c6a79dbba62ab80f40d656ca9.

bdach committed:
Remove more no-longer-required checks

The scenario that remaining guard was trying to protect against is
obviated by and no longer possible after
776fabd77c5a6225e69a0b14e79b9e097692320c.

peppy commented:
Not 100% about the filename removal. Stable still uses this as a fallback in order to let users update to the latest version in-game. Thoughts?

bdach commented:
My thoughts are that online ID in the beatmap file and the beatmap filename are similarly reliable, i.e. not at all, and if we're bringing back one, we may as well bring both back, and just rely on md5 checks at submission time to prevent funny business.

Just depends on how many layers of safety you're comfortable with I guess. I am not sure how to conclusively determine whether excluding filename is going to be a problem or not.

peppy commented:
My thoughts are that it hasn't ever presented an issue to date in stable, so we could probably still keep it around. I think it's the only way we can recover the case where a user downloads a beatmap with `.osu` files that are outdated (ie. MD5 doesn't match database hash). This can regularly happen in cases like:

- Beatmap pack archives (aren't updated when rare fixes are made to ranked maps)
- Mirrors incorrectly having outdated version of beatmaps (brought up in recent threads)

bdach commented:
I don't have willpower to fight on that point, but you do realise I will have to revert https://github.com/ppy/osu/pull/30453/commits/2b0fd3558f2f7f1c6a79dbba62ab80f40d656ca9 etc and bring the hash checks back if I'm to allow filename again?

peppy commented:
> Unless you're again saying that that's fine and that the onus of checking the hash should be on the score submission process?

I think this is valid thinking, yeah. After all, a user could just edit things in realm post-import. I feel we were doing far too much on import to try and get things in a good state, and probably need to fall back to "what worked in stable" for now

bdach committed:
Bring back matching by filename when performing online metadata lookups

bdach commented:
Alright, filename lookup restored with no guardrails, online ID still excluded.

I have tested that the submission process correctly detects this and prevents score token issuance when attempting to play a doctored beatmap.
